File: tailor-cloud-app/lib/models/order.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'order.freezed.dart';
part 'order.g.dart';

/// 注文ステータス
enum OrderStatus {
  @JsonValue('Draft')
  draft,
  @JsonValue('Confirmed')
  confirmed,
  @JsonValue('InProduction')
  inProduction,
  @JsonValue('Completed')
  completed,
  @JsonValue('Cancelled')
  cancelled,
}

/// 注文詳細情報
@freezed
class OrderDetails with _$OrderDetails {
  const factory OrderDetails({
    String? description,
    @JsonKey(name: 'measurement_data') Map<String, dynamic>? measurementData,
    Map<String, dynamic>? adjustments,
  }) = _OrderDetails;

  factory OrderDetails.fromJson(Map<String, dynamic> json) =>
      _$OrderDetailsFromJson(json);
}

/// 注文モデル
@freezed
class Order with _$Order {
  const factory Order({
    required String id,
    @JsonKey(name: 'tenant_id') required String tenantId,
    @JsonKey(name: 'customer_id') required String customerId,
    @JsonKey(name: 'fabric_id') required String fabricId,
    required OrderStatus status,
    @JsonKey(name: 'total_amount') required int totalAmount,
    @JsonKey(name: 'payment_due_date') required DateTime paymentDueDate,
    @JsonKey(name: 'delivery_date') required DateTime deliveryDate,
    OrderDetails? details,
    @JsonKey(name: 'compliance_doc_url') String? complianceDocUrl,
    @JsonKey(name: 'compliance_doc_hash') String? complianceDocHash,
    @JsonKey(name: 'created_at') required DateTime createdAt,
    @JsonKey(name: 'updated_at') required DateTime updatedAt,
    @JsonKey(name: 'created_by') required String createdBy,
  }) = _Order;

  factory Order.fromJson(Map<String, dynamic> json) => _$OrderFromJson(json);
}

/// 注文拡張メソッド
extension OrderExtension on Order {
  /// ステータスの表示ラベル
  String get statusLabel {
    switch (status) {
      case OrderStatus.draft:
        return '下書き';
      case OrderStatus.confirmed:
        return '確定';
      case OrderStatus.inProduction:
        return '製作中';
      case OrderStatus.completed:
        return '完了';
      case OrderStatus.cancelled:
        return 'キャンセル';
    }
  }

  /// 金額の表示（円）
  String get amountDisplay {
    return '¥${totalAmount.toString().replaceAllMapped(
          RegExp(r'(\d)(?=(\d{3})+(?!\d))'),
          (Match m) => '${m[1]},',
        )}';
  }
}

/// 注文作成リクエスト
@freezed
class CreateOrderRequest with _$CreateOrderRequest {
  const factory CreateOrderRequest({
    @JsonKey(name: 'customer_id') required String customerId,
    @JsonKey(name: 'fabric_id') required String fabricId,
    @JsonKey(name: 'total_amount') required int totalAmount,
    @JsonKey(name: 'delivery_date') required DateTime deliveryDate,
    OrderDetails? details,
  }) = _CreateOrderRequest;

  factory CreateOrderRequest.fromJson(Map<String, dynamic> json) =>
      _$CreateOrderRequestFromJson(json);
}

/// 注文確定リクエスト
@freezed
class ConfirmOrderRequest with _$ConfirmOrderRequest {
  const factory ConfirmOrderRequest({
    @JsonKey(name: 'order_id') required String orderId,
    @JsonKey(name: 'principal_name') required String principalName,
  }) = _ConfirmOrderRequest;

  factory ConfirmOrderRequest.fromJson(Map<String, dynamic> json) =>
      _$ConfirmOrderRequestFromJson(json);
}

File: suit-mbti-web-app/src/types/order.ts
// 注文およびコンプライアンス文書まわりの型定義

export type OrderStatus = 'Draft' | 'Confirmed' | 'Cancelled' | 'Completed' | string;

export interface Order {
  id: string;
  tenant_id: string;
  customer_id: string;
  fabric_id: string;
  status: OrderStatus;
  total_amount: number;
  tax_excluded_amount?: number;
  tax_amount?: number;
  tax_rate?: number;
  payment_due_date?: string;
  delivery_date: string;
  details?: {
    description?: string;
    measurement_data?: Record<string, unknown>;
    adjustments?: Record<string, unknown>;
  };
  created_at: string;
  updated_at: string;
  created_by?: string;
}

export interface CreateOrderRequest {
  customer_id: string;
  fabric_id: string; // バックエンドでは必須
  total_amount: number;
  delivery_date: string; // ISO 8601 (RFC3339形式: "2025-12-31T00:00:00Z")
  details: {
    description: string;
    measurement_data?: Record<string, unknown>;
    adjustments?: Record<string, unknown>;
  };
  // バックエンドで自動設定されるが、開発環境では明示的に指定可能
  tenant_id?: string;
  created_by?: string;
}

export interface GenerateDocumentResponse {
  order_id: string;
  doc_url: string;
  doc_hash: string;
  generated_at: string;
}


